### PR TITLE
feat(Interactions): add modal and text input interactions

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

**Reference Discord API Docs PRs or commits:**
- https://github.com/discord/discord-api-docs/pull/4459

### CODE REVIEW

File: payloads/v9/channel.ts [hunk 1]
```diff
@@ -1207,7 +1217,35 @@ export interface APISelectMenuOption {
...
+
 /**
  * https://discord.com/developers/docs/interactions/message-components#message-components
  */
-export type APIMessageComponent = APIActionRowComponent | APIButtonComponent | APISelectMenuComponent;
+export type APIMessageComponent =
+	| APIActionRowComponent
+	| APIButtonComponent
+	| APISelectMenuComponent
+	| APIInputTextComponent;
```
Comment (vladfrangu) [inline payloads/v9/channel.ts:1251]: Well you can't use this in Messages so its not a MessageComponent :D
Probably time to define a second ActionRow for modals? Idk..

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: payloads/v9/channel.ts (hunk 1 above)
Comment (suneettipirneni) [reply payloads/v9/channel.ts:1251]: Most likely

File: payloads/v8/channel.ts [hunk 2]
```diff
@@ -949,6 +953,13 @@ export interface APIActionRowComponent extends APIBaseMessageComponent<Component
 	components: Exclude<APIMessageComponent, APIActionRowComponent>[];
 }
 
+export interface APIModalActionRowComponent extends APIBaseMessageComponent<ComponentType.ActionRow> {
+	/**
+	 * The components in the ActionRow
+	 */
+	components: Exclude<APIModalComponent, APIActionRowComponent>[];
```
Comment (vladfrangu) [inline payloads/v8/channel.ts:960]:
```suggestion
	components: Exclude<APIModalComponent, APIModalActionRowComponent>[];
```

File: payloads/v8/channel.ts [hunk 3]
```diff
@@ -1078,7 +1095,33 @@ export interface APISelectMenuOption {
...
+	 */
+	max_length?: number;
+}
+
 /**
  * https://discord.com/developers/docs/interactions/message-components#message-components
  */
 export type APIMessageComponent = APIActionRowComponent | APIButtonComponent | APISelectMenuComponent;
+
+export type APIModalComponent = APIActionRowComponent | APIInputTextComponent;
```
Comment (vladfrangu) [inline payloads/v8/channel.ts:1127]:
```suggestion
export type APIModalComponent = APIModalActionRowComponent | APIInputTextComponent;
```

File: payloads/v9/channel.ts [hunk 4]
```diff
@@ -1078,6 +1082,13 @@ export interface APIActionRowComponent extends APIBaseMessageComponent<Component
 	components: Exclude<APIMessageComponent, APIActionRowComponent>[];
 }
 
+export interface APIModalActionRowComponent extends APIBaseMessageComponent<ComponentType.ActionRow> {
+	/**
+	 * The components in the ActionRow
+	 */
+	components: Exclude<APIModalComponent, APIActionRowComponent>[];
```
Comment (vladfrangu) [inline payloads/v9/channel.ts:1089]:
```suggestion
	components: Exclude<APIModalComponent, APIModalActionRowComponent>[];
```

File: payloads/v9/channel.ts [hunk 5]
```diff
@@ -1207,7 +1224,33 @@ export interface APISelectMenuOption {
...
+	 */
+	max_length?: number;
+}
+
 /**
  * https://discord.com/developers/docs/interactions/message-components#message-components
  */
 export type APIMessageComponent = APIActionRowComponent | APIButtonComponent | APISelectMenuComponent;
+
+export type APIModalComponent = APIActionRowComponent | APIInputTextComponent;
```
Comment (vladfrangu) [inline payloads/v9/channel.ts:1256]:
```suggestion
export type APIModalComponent = APIModalActionRowComponent | APIInputTextComponent;
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Comment by suneettipirneni** [comment]
Upstream commit documentation landed

**Comment by vladfrangu** [comment]
"landed"

**Comment by suneettipirneni** [comment]
> "landed"

I can't read

File: payloads/v8/channel.ts [hunk 6]
```diff
@@ -1140,7 +1157,45 @@ export interface APISelectMenuOption {
 	default?: boolean;
 }
 
+// TODO: Add api links
+export interface APIInputTextComponent extends APIBaseMessageComponent<ComponentType.InputText> {
```
Comment (vladfrangu) [inline payloads/v8/channel.ts:1161]: Its TextInput not InputText

**Comment by DTrombett** [comment]
New upstream PR: discord/discord-api-docs#4459 🎉

**Comment by SuperchupuDev** [comment]
upstream pr was merged :tada:

File: deno/payloads/v8/_interactions/modalSubmit.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import type { APIBaseInteraction, InteractionType, APIModalActionRowComponent } from '../mod.ts';
+
+/**
+ * https://discord.com/developers/docs/interactions/message-components#component-object-component-structure
```
Comment (vladfrangu) [inline deno/payloads/v8/_interactions/modalSubmit.ts:4]: make sure the link is correct

File: deno/payloads/v8/_interactions/responses.ts [hunk 8]
```diff
@@ -92,3 +103,9 @@ export type APIInteractionResponseCallbackData = Omit<
 export interface APICommandAutocompleteInteractionResponseCallbackData {
 	choices?: APIApplicationCommandOptionChoice[];
 }
+
+export interface APIModalInteractionResponseCallbackData {
+	custom_id: string;
```
Comment (vladfrangu) [inline deno/payloads/v8/_interactions/responses.ts:108]: typedocs

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: deno/payloads/v8/_interactions/modalSubmit.ts (hunk 7 above)
Comment (suneettipirneni) [reply deno/payloads/v8/_interactions/modalSubmit.ts:4]: Doesn't seem to be any specific docs for this so I'll remove it

File: deno/payloads/v8/_interactions/modalSubmit.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+export interface APIModalSubmission {
+	/**
+	 * A developer-defined identifier for the component, max 100 characters
+	 */
+	custom_id: string;
+	/**
+	 * A list of child components
+	 */
+	components?: APIModalActionRowComponent[];
```
Comment (vladfrangu) [inline deno/payloads/v8/_interactions/modalSubmit.ts:11]: Wait, is this what we get? a row of components? That doesn't seem right...how do we get the data from the users 👀

File: deno/payloads/v8/channel.ts [hunk 10]
```diff
@@ -1011,6 +1015,19 @@ export interface APIActionRowComponent extends APIBaseMessageComponent<Component
 	components: Exclude<APIMessageComponent, APIActionRowComponent>[];
 }
 
+export interface ModalSubmitComponent {
```
Comment (vladfrangu) [inline deno/payloads/v8/channel.ts:1018]: This should go in modalSubmit, not here. By making modal action row use this interface, the types are wrong.

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by kyranet — APPROVED** [review]

File: deno/payloads/v8/channel.ts [hunk 11]
```diff
@@ -1140,7 +1153,52 @@ export interface APISelectMenuOption {
...
+	 */
+	required?: boolean;
+}
+
 /**
  * https://discord.com/developers/docs/interactions/message-components#message-components
  */
-export type APIMessageComponent = APIActionRowComponent | APIButtonComponent | APISelectMenuComponent;
+export type APIMessageComponent =
+	| APIActionRowComponent<APIMessageComponent | APIModalComponent>
```
Comment (vladfrangu) [inline deno/payloads/v8/channel.ts:1198]: Thats not correct, message components cannot have modal ones in them

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**